Keep-series images: retain burst empties + selectable trigger species; harden long species selection lists

Daveiano opened the pull request:
Hello there, 

I am working in wolf conservation in Germany, and we are working a lot with burst shots. For our use case, we are keeping all images of a series if one image contains a detection, so we also keep the empties of this burst. This PR adds 3 new configs for post:  `var_keep_series` (boolean), `var_keep_series_species` (array), and `var_keep_series_seconds` (float) to configure which species need this handling (we only want this for wolf) and to configure the timing of the burst shots. I found that there is no relevant metadata in the images to know if they belong to the same series, except for the date and time, so we can now configure the interval in seconds for an image to be considered from the same series of images.

Tbh, I extensively used AI to do the job. AddaxAI_GUI.py is a large collection of logic, and it takes some time to understand, so I used AI to get on the right track. Please keep that in mind when/if reviewing this. I tested this for a few cards full of images and so far it's working realy well.

### Summary

This PR adds an optional “keep series images” post-processing behavior: when an animal is detected in one image, AddaxAI can also keep the empty images that belong to the same camera-trap burst/series (within a configurable time window). It also exposes a selectable “trigger species” list (defaults to “any animal detection”).

Additionally, this PR updates SpeciesSelectionFrame to safely render very long checkbox lists (e.g., aggregated species across models) without overlapping items near the end of the scroll area.

### User-facing changes

- New post-processing option: “Keep series (if animal detected in one image)”
- New sub-options (shown when enabled):
  - Seconds (var_keep_series_seconds) — time window for defining a “series”
  - Select species… — opens a modal to choose trigger classes (var_keep_series_species)
- When no trigger species are selected, behavior defaults to any animal-like detection (excluding person/vehicle) triggering “keep the series”.

### Technical details (for faster review)

- Post-processing series logic
  - Adds helpers:
    - build_image_timestamp_index(src_dir, file_list) — caches timestamps per file
    - find_series_images(target_rel_path, timestamp_index, window_seconds=10, require_same_camera=True) — returns the burst neighbors
  - Integrates into the separation/move step using:
    - keep_series (checkbox)
    - keep_series_seconds (slider)
    - keep_series_species = global_vars.get('var_keep_series_species', [])
    - already_moved_files set to avoid double-moving/copying files
- Trigger species aggregation across all supported models
  - New helper: get_all_supported_model_classes(force_refresh=False)
  - Collects the union of species/class labels from all supported classifier models, caches into _ALL_SUPPORTED_MODEL_CLASSES_CACHE, and returns a sorted list.
  - This allows selecting species like "wolf" even if not every model supports that class.
- New modal for keep-series trigger selection
  - open_keep_series_species_selection() (stores selection to global_vars['var_keep_series_species'] and persists via write_global_vars({'var_keep_series_species': chosen}))
  - Small UI counter updates via dsp_keep_series_species (“Any” or count)
- Why SpeciesSelectionFrame changed
  - SpeciesSelectionFrame(customtkinter.CTkScrollableFrame) now supports auto multi-column layouts to prevent extremely tall single-column lists from triggering canvas rendering issues (checkboxes overlapping / last items not visible). See screenshot.
  - Implementation:
    - Auto layout constants: AUTO_MAX_ROWS_PER_COLUMN, AUTO_MAX_COLUMNS
    - n_columns=None defaults to auto via _resolve_n_columns(len(all_classes))
    - add_item() grids items by (row = idx // n_columns, col = idx % n_columns)

<img width="875" height="742" alt="Screenshot 2025-12-25 235928" src="https://github.com/user-attachments/assets/561fe5c9-1167-4e3b-8fa2-6f71f207afb3" />

### Related TODOs (already in repo)

This feature is closely related to:

- [TODO: SEQ SEP](https://github.com/Daveiano/AddaxAI/blob/main/AddaxAI_GUI.py#L22) – “feature to separate images into sequence subdirs / include all images in the sequence”
- [TODO: LOG SEQUENCE INFO](https://github.com/Daveiano/AddaxAI/blob/main/AddaxAI_GUI.py#L21) – “add sequence information to JSON/CSV/XLSX”
- [TODO: EMPTIES](https://github.com/Daveiano/AddaxAI/blob/main/AddaxAI_GUI.py#L20) – “option to skip empties from being copied” (this PR keeps empties selectively when part of a triggered series)

### How to test

1. Use a dataset with bursts (4–5 images per trigger).
2. Run detection/classification as usual.
3. In post-processing:
  - enable “Keep series…”
  - set Seconds to cover your burst spacing (e.g., 2–5s)
  - optionally select a trigger species (e.g., wolf)
4. Confirm that if one image in a burst contains the trigger, the empty neighbor images in that burst are also kept/moved.

### Backwards compatibility

- Feature is opt-in (var_keep_series defaults to False).
- If var_keep_series_species is empty, behavior matches “trigger on any animal detection”.

PetervanLunteren commented:
Hi @Daveiano, sorry for the long wait! 

I gave it a go and there were no errors, but I did find the following:

1. When I click the "select species" it opens a window with all model classes, not just the one that the user selected, which is a bit of an overkill. Could you adjust that? See attached. 
2. I noticed that the checbox lists are all adjusted in padding/margins. Not a huge deal, but if you could restore that to the orignal is would be great.  (see attached)
3. When I tried to run it on bursts and exported to XLSX I did not get an extra column. (see attached) Isn't that what is expected? Or does it only affect the "Separate files into subdirectories"? If so, could you add it to the "separation options" subframe so it is clear it only affects the file separation?
4. Not a huge deal, but it doesnt follow the greyed out pattern the rest folllows if step 4 is greyed out (see attached). 

Thanks!
<img width="798" height="708" alt="Screenshot 2026-01-06 at 16 38 54" src="https://github.com/user-attachments/assets/e910a4ae-1420-4fad-b36d-3d050125b94a" />
<img width="742" height="836" alt="Screenshot 2026-01-06 at 16 42 51" src="https://github.com/user-attachments/assets/483f6178-dc5d-4c2f-bd8c-8b47c5fd1a92" />
<img width="1306" height="952" alt="Screenshot 2026-01-06 at 16 44 10" src="https://github.com/user-attachments/assets/7c1bb4d8-be7c-41df-8320-de0b8d86d27e" />
[results.xlsx](https://github.com/user-attachments/files/24455558/results.xlsx)

Daveiano commented:
Hi @PetervanLunteren, no need to apologise for the long wait. I also had some things to do, but I managed to take a look at your comment!

> When I click the "select species" it opens a window with all model classes, not just the one that the user selected, which is a bit of an overkill. Could you adjust that? See attached.

The select species modal of the keep series option now only shows the classes of the current selected model. However, it will save the value globally, so if you switch models, the changes persist. I also included a clear button which will reset the selected species to allow the handling to "any" detections which is also shown in the UI.

> I noticed that the checbox lists are all adjusted in padding/margins. Not a huge deal, but if you could restore that to the orignal is would be great. (see attached)

Since we are not displaying the long list of all classes of all models, the original SpeciesSelectionFrame works again, so I switched back to the original code, which should also restore the original paddings and margins.

> When I tried to run it on bursts and exported to XLSX I did not get an extra column. (see attached) Isn't that what is expected? Or does it only affect the "Separate files into subdirectories"? If so, could you add it to the "separation options" subframe so it is clear it only affects the file separation?

I simply moved everything inside the separation options subframe to make it clear. Guess updating the generated XLSX would be a seperate pull request. Just out of curiosity, what would you expect for the column to contain? One of these?

- a series/burst identifier (sequence_id, burst_id, series_index, etc.)
- a boolean like “kept_by_series” / “in_series_of_trigger”
- maybe the time-window grouping value that put images into a series

> Not a huge deal, but it doesnt follow the greyed out pattern the rest folllows if step 4 is greyed out (see attached).

Fixed.

I pushed all changes to the [original branch of the PR](https://github.com/Daveiano/AddaxAI/tree/main), but since it is already merged, it won't update anymore. Should I open a new PR?
